deepreg/model/layer.py

```diff
@@ -665,3 +668,204 @@ def call(self, inputs, training=None, **kwargs):
         r2 = self._conv3d_block(inputs=h0, training=training)
         h1 = self._residual_block(inputs=[r2, r1], training=training)
         return h1
+
+
+class BSplines3DTransform(tf.keras.layers.Layer):
+    """
+     Layer for BSplines interpolation with precomputed cubic spline filters.
+     It assumes a full sized image from which: (1) it compute the contol points values by downsampling the initial
+     image (2) performs the interpolation and (3) crops the image around the valid values.
+
+    :param cp_spacing: int or tuple of three ints specifying the spacing (in pixels) in each dimension.
+                      When a single int is used, the same spacing to all dimensions is used
+    :param kwargs:
+    """
+
+    def __init__(self, cp_spacing: (int, tuple), **kwargs):
+
+        super(BSplines3DTransform, self).__init__(**kwargs)
+
+        self.filters = []
+
+        if isinstance(cp_spacing, int):
+            self.cp_spacing = (cp_spacing, cp_spacing, cp_spacing)
+        else:
+            self.cp_spacing = cp_spacing
+
+    def build(self, input_shape: tuple):
+        """
+        :param input_shape: tuple with the input shape
+        :return:
+        """
+
+        super(BSplines3DTransform, self).build(input_shape=input_shape)
+
+        b = {
+            0: lambda u: np.float32((1 - u) ** 3 / 6),
+            1: lambda u: np.float32((3 * (u ** 3) - 6 * (u ** 2) + 4) / 6),
+            2: lambda u: np.float32((-3 * (u ** 3) + 3 * (u ** 2) + 3 * u + 1) / 6),
+            3: lambda u: np.float32(u ** 3 / 6),
+        }
+
+        filters = np.zeros(

```

Comment:
I'm not sure I can understand this by reading the code, is there any docs as ref? I'd be happy to read it and double check the code. For now it's kind of like LGTM lol

btw do you think it will be possible to isolate the code from `filters = np.zeros` to the line before `self.filter = tf.convert_to_tensor(filters)`? so that we can test the values not just the shapes.

Reply:
or maybe what we can do for testing the value, is to compare the results of current code with the results of your previous code in test.

so in test, we call this more efficient code, and compare with the easier to understand code, if their results are the same, it validates your efficient code.

Reply:
https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=796284 is this the reference?

Reply:
Yes, I compared it with the commented code, which is more "literal" but less efficient and it gives the same results.

Regarding the implementation yes, the reference is correct
![comparison](https://user-images.githubusercontent.com/19728301/101179786-fc1a6480-364a-11eb-9d24-42118a3439ac.png)